### DISCUSSION TITLE
Is there a way to define a customized Line Style such as "LinePixels"

### DISCUSSION THREAD

**Discussion by NguyenBentley** [iTwin/itwinjs-core#680] [Q&A] [answered]
iModelJs defines some line styles in  export declare enum LinePixels {…} that can be used out of the box
 
My question is: Is there a way to define a customized Line Style "LinePixels" ?
The way via “createStrokePatternComponent" is a lot complicated and not conform since the StrokeProps.length is in meter
 
I am happy if you can give me some hint or example

**Comment by pmconne** [comment]
There is currently no way to define your own `LinePixels` pattern. What end-result are you trying to achieve?

**Comment by NguyenBentley** [comment]
Such as [LinePixels.Code1, LinePixels.Code2, LinePixels.Code3, LinePixels.Code4, LinePixels.Code5, LinePixels.Code6, LinePixels.Code7]; but e.g. LinePixels.Code1 that means "1 lit pixel followed by 7 unlit pixels" i want LinePixels.CodeNLit_MUnlit ""N lit pixel followed by M unlit pixels" totally customized

**Comment by pmconne** [comment]
Yes, I am asking "for what purpose"? *Why* do you want this?

**Comment by NguyenBentley** [comment]
We do import from DEXPI or autocad and want to draw the lines in iModelJs exactly as the custom lines

**Reply by pmconne** [reply]
Thanks. Can you show me some examples, and/or link to documentation about how custom line patterns are defined in AutoCAD and/or DEXPI? If they can be expressed as a 32-bit integer (bitmask) we could probably make them work (WebGL 2 only).

**Comment by NguyenBentley** [comment]
e.g.
export enum StrokeDashArray {
  Dashed_1_1 = "1 1",
  Dashed_2_1 = "2 1",
  Dashed_3_1 = "3 1",
  Dashed_5_1 = "5 1",

  Solid = "",
  Dashed_1_7 = "1 7",
  Dashed_5_3 = "5 3",
  Dashed_11_5 = "11 5",
  Dashed_7_4_1_1 = "7 4 1 1",
  Dashed_3_5 = "3 5",
  Dashed_5_3_1_3_1_3 = "5 3 1 3 1 3",
  Dashed_8_3_2_3 = "8 3 2 3",
  Dashed_2_2 = "2 2",
  Dashed_1_31 = "1 31",
}

**Reply by pmconne** [reply]
I'm asking how a user would describe the pattern to AutoCAD.

**Comment by NguyenBentley** [comment, marked as answer]
We convert the dash lines of the autocad & DEXPI to these StrokeDashArray and want to draw them the same in iModelJs Design Review. The definition of StrokeDashArray, as you know, can be extended & modified in the future

**Reply by donhfu** [reply]
It sounds to me that you intend to convert a unit line style to a unitless line style.  You may get more issues than help as these are two different concepts.  A line type in ACAD is always in true units.  When you zoom in and out of the view, the dashes and gaps change size.  You can even measure the distance between two components.  If change them to unitless, they will look different than they are in ACAD.  My recommendation is to stay with custom linestyles.

**Reply by bbastings** [reply]
Here are some examples of how to define a line style:

https://github.com/imodeljs/imodeljs/blob/master/core/backend/src/test/standalone/GeometryStream.test.ts

There isn't a simple dash/gap example, but "create GeometricElement3d using compound style with dash widths and symbol" does show how to create the stroke pattern component.

Basically this:

    const lsStrokes: LineStyleDefinition.Strokes = [];
    lsStrokes.push({ length: 0.25, strokeMode: LineStyleDefinition.StrokeMode.Dash });
    lsStrokes.push({ length: 0.1 });

    const strokePatternData = LineStyleDefinition.Utils.createStrokePatternComponent(imodel, { descr: "TestDashGapStrokePattern", strokes: lsStrokes });
    assert.isTrue(undefined !== strokePatternData);

    const styleId = LineStyleDefinition.Utils.createStyle(imodel, IModel.dictionaryId, "TestDashGapStyle", strokePatternData!);
    assert.isTrue(Id64.isValidId64(styleId));

    const builder = new GeometryStreamBuilder();
    const params = new GeometryParams(categoryId);

    params.styleInfo = new LineStyle.Info(styleId);
    builder.appendGeometryParamsChange(params);
    builder.appendGeometry(LineSegment3d.create(Point3d.createZero(), Point3d.create(5, 5, 0)));

    const elementProps = createPhysicalElementProps(seedElement, undefined, builder.geometryStream);
    const newId = imodel.elements.insertElement(elementProps);

**Reply by NguyenBentley** [reply]
Thank
I will try it
